core/src/main/java/org/apache/carbondata/core/scan/expression/geo/PolygonExpression.java

```diff
@@ -0,0 +1,117 @@
+
+import java.util.List;
+
+import org.apache.carbondata.common.annotations.InterfaceAudience;
+import org.apache.carbondata.core.metadata.datatype.DataTypes;
+import org.apache.carbondata.core.scan.expression.ColumnExpression;
+import org.apache.carbondata.core.scan.expression.Expression;
+import org.apache.carbondata.core.scan.expression.ExpressionResult;
+import org.apache.carbondata.core.scan.expression.LiteralExpression;
+import org.apache.carbondata.core.scan.expression.conditional.GreaterThanEqualToExpression;
+import org.apache.carbondata.core.scan.expression.conditional.LessThanEqualToExpression;
+import org.apache.carbondata.core.scan.expression.exception.FilterIllegalMemberException;
+import org.apache.carbondata.core.scan.expression.exception.FilterUnsupportedException;
+import org.apache.carbondata.core.scan.expression.logical.AndExpression;
+import org.apache.carbondata.core.scan.expression.logical.OrExpression;
+import org.apache.carbondata.core.scan.expression.logical.RangeExpression;
+import org.apache.carbondata.core.scan.expression.logical.TrueExpression;
+import org.apache.carbondata.core.scan.filter.intf.ExpressionType;
+import org.apache.carbondata.core.scan.filter.intf.RowIntf;
+import org.apache.carbondata.core.util.CustomIndex;
+
+/**
+ * InPolygon expression processor. It inputs the InPolygon string to the GeoHash implementation's
+ * query method, gets the list of ranges of GeoHash IDs to filter as an output. And then, multiple
+ * range expressions are build from those list of ranges.
+ */
+@InterfaceAudience.Internal
+public class PolygonExpression extends Expression {
+  private String polygon;
+  private String columnName;
+  private CustomIndex<Long, String, List<Long[]>> handler;
+  private List<Long[]> ranges;
+
+  public PolygonExpression(String polygon, String columnName, CustomIndex handler) {
+    this.polygon = polygon;
+    this.handler = handler;
+    this.columnName = columnName;
+  }
+
+  /**
+   * This method builds the GeoHash range expressions from the list of ranges of GeoHash IDs.
+   */
+  public void buildRangeExpression() {
+    try {
+      ranges = handler.query(polygon);
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+
+    // Convert these ranges into range expressions
+    Expression expression = null;
+    Expression prevExpression = null;
+    Expression rangeExpression;
+    for (Long[] range : ranges) {
+      assert (range.length == 2);
+      rangeExpression = new RangeExpression(
+              new LessThanEqualToExpression(
+                      new ColumnExpression(columnName, DataTypes.LONG),
+                      new LiteralExpression(range[1], DataTypes.LONG)),
+              new GreaterThanEqualToExpression(
+                      new ColumnExpression(columnName, DataTypes.LONG),
+                      new LiteralExpression(range[0], DataTypes.LONG)));
+      expression = new AndExpression(rangeExpression, new TrueExpression(null));
```

Comment:
AndExpression(true, range) = range itself. I see other places also we use like this, but may be some one can explain why like this ?
@ravipesala, @kumarvishal09

Reply:
@ajantha-bhat : It was added as optimization for a <= x && a >= y case. 
If column 'a' is sorted, we can convert above filter to Range(x-y) && (true). Hence it was added. 
Here I guess it is not required. @VenuReddy2103 can check about this.

Reply:
Modified